vendor/github.com/coreos/go-oidc/v3/oidc/oidc.go

```diff
@@ -55,6 +58,36 @@ func ClientContext(ctx context.Context, client *http.Client) context.Context {
 	return context.WithValue(ctx, oauth2.HTTPClient, client)
 }
 
+// cloneContext copies a context's bag-of-values into a new context that isn't
+// associated with its cancellation. This is used to initialize remote keys sets
+// which run in the background and aren't associated with the initial context.
+func cloneContext(ctx context.Context) context.Context {
+	cp := context.Background()
+	if c, ok := ctx.Value(oauth2.HTTPClient).(*http.Client); ok {
+		cp = ClientContext(cp, c)
+	}
+	return cp
+}
+
+// InsecureIssuerURLContext allows discovery to work when the issuer_url reported
+// by upstream is mismatched with the discovery URL. This is meant for integration
+// with off-spec providers such as Azure.
+//
+//    discoveryBaseURL := "https://login.microsoftonline.com/organizations/v2.0"
+//    issuerURL := "https://login.microsoftonline.com/my-tenantid/v2.0"
+//
+//    ctx := oidc.InsecureIssuerURLContext(parentContext, issuerURL)
+//
+//    // Provider will be discovered with the discoveryBaseURL, but use issuerURL
+//    // for future issuer validation.
+//    provider, err := oidc.NewProvider(ctx, discoveryBaseURL)
+//
+// This is insecure because validating the correct issuer is critical for multi-tenant
+// proivders. Any overrides here MUST be carefully reviewed.
+func InsecureIssuerURLContext(ctx context.Context, issuerURL string) context.Context {
```

Comment:
this is unused currently, right?

Reply:
right! i'll wire this up in the next PR with the addition of `discoveryURL` in the config.

Reply:
it's a terrifying name to have in our call stack, just trying to understand exactly what it does and how it works

Reply:
It's only used [here](https://github.com/kubernetes/kubernetes/pull/123402/files#diff-06e9f00b62aa58d544d62f2d046009446a72e681c9f26450b15fafcb0ffd8ce6R168-R172)


```go
ctx := oidc.InsecureIssuerURLContext(parentContext, opts.JWTAuthenticator.Issuer.URL)
provider, err := oidc.NewProvider(ctx, opts.JWTAuthenticator.Issuer.DiscoveryURL) // if discovery url is not empty
...
```